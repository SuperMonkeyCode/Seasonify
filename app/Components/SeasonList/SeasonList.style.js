export default {
  seasonHeader: {
    flexDirection: 'column',
    justifyContent: 'flex-start',
    alignItems: 'flex-end', // not-working, so using alignSelf
  },
  seasonTitle: {
    alignSelf: 'flex-end',
  },
  seasonInfos: {
    alignSelf: 'flex-end',
  },
  viewerList: {
    flexDirection: 'row',
    flexWrap: 'wrap',
    justifyContent: 'flex-start',
    alignItems: 'flex-start', // not-working, so using alignSelf
    paddingHorizontal: 20,
  },
  viewer: {
    alignSelf: 'flex-start',
    marginHorizontal: 2,
    marginVertical: 2,
  },
};
