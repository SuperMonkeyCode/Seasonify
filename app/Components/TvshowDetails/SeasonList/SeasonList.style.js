export default {
  seasonHeader: {
    flexDirection: 'row',
    flexWrap: 'wrap',
    justifyContent: 'space-between',
    alignItems: 'flex-start',
  },
  seasonButton: {},
  seasonTitle: {
    flexGrow: 1,
    textAlign: 'right',
  },
  seasonInfos: {
    flexBasis: '100%',
    textAlign: 'right',
  },
  // icon: {
  //   fontSize: 20,
  // },
  viewerList: {
    flexDirection: 'row',
    flexWrap: 'wrap',
    justifyContent: 'flex-start',
    alignItems: 'flex-start',
    paddingHorizontal: 20,
  },
  viewer: {
    marginHorizontal: 2,
    marginVertical: 2,
  },
};
