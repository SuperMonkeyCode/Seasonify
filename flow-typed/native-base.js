/* @flow */
declare module 'native-base' {
  declare var exports: any;
}
