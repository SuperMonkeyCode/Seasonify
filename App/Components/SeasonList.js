import React from 'react';
import { List, ListItem, Text, Grid, Col, Badge } from 'native-base';

import { Metrics, AtomicStyles } from '../Themes';

const SeasonList = ({ seasons }) => (
  <List>
    {seasons &&
      Object.keys(seasons).reverse().map(id => (
        <ListItem key={id}>
          <Grid>
            <Col size={Metrics.columnLeft} style={AtomicStyles.noFlex}>
              <Badge info>
                <Text>{`Season ${id}`}</Text>
              </Badge>
              <Text note>
                {seasons[id].episodes && `${seasons[id].episodes} ep. `}
                {seasons[id].year && `in ${seasons[id].year}`}
              </Text>
            </Col>
            <Col size={Metrics.columnRight}>
              {/* <Text>Renaud, Sylvain, Papa&Maman</Text> */}
            </Col>
          </Grid>
        </ListItem>
      ))}
  </List>
);

export default SeasonList;
