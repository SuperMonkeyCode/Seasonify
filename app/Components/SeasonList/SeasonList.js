/* @flow */
import { Button, Col, Grid, List, ListItem, Text } from 'native-base';
import { compose, pure, withHandlers } from 'recompose';
import React from 'react';

import { Metrics } from 'Themes';
import type { Seasons, ViewerInfo } from 'Types';
import { getContrastingTextColor } from 'Libs/Helpers';

import styles from './SeasonList.style';

type Props = {
  /* parent */
  seasons: Seasons,
  navigation: Object,
  /* connect */
  seasonViewings: { [season: string]: ViewerInfo[] },
  /* HOC */
  handleViewerPress: Function,
};

const enhance = compose(
  pure,
  withHandlers({
    handleViewerPress: ({ navigation }: Props) => (viewer: ViewerInfo) => {
      navigation.navigate('FriendDetailsPage', {
        friendId: viewer.friendId,
        friendName: viewer.name,
      });
    },
  }),
);

const getBadgeBgColor = (color: string) => ({ backgroundColor: color });

const getBadgeColor = (color: string) => ({ color: getContrastingTextColor(color) });

function SeasonList({ seasons, seasonViewings, handleViewerPress }: Props) {
  return (
    <List>
      {seasons &&
        objectValues(seasons).reverse().map(season =>
          (<ListItem key={season.id}>
            <Grid>
              <Col size={Metrics.columnLeft} style={styles.seasonHeader}>
                <Text style={styles.seasonTitle}>{`Season ${season.id}`}</Text>
                <Text note style={styles.seasonInfos}>
                  {season.episodes && `${season.episodes} ep.`}
                  {season.year && ` in ${season.year}`}
                </Text>
              </Col>
              <Col size={Metrics.columnRight} style={styles.viewerList}>
                {seasonViewings[season.id] &&
                  seasonViewings[season.id].map(viewer =>
                    (<Button
                      small
                      key={viewer.friendId}
                      style={{ ...styles.viewer, ...getBadgeBgColor(viewer.color) }}
                      onPress={() => handleViewerPress(viewer)}
                    >
                      <Text style={getBadgeColor(viewer.color)}>
                        {viewer.name}
                      </Text>
                    </Button>),
                  )}
              </Col>
            </Grid>
          </ListItem>),
        )}
    </List>
  );
}

export default enhance(SeasonList);
