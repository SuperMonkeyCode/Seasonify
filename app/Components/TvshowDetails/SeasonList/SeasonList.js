/* @flow */
import { Button, Col, Grid, Icon, Left, List, ListItem, Text, View } from 'native-base';
import { TouchableOpacity } from 'react-native';
import { compose, pure, withHandlers } from 'recompose';
import React from 'react';

import { Metrics } from 'Themes';
import { getContrastingTextColor } from 'Libs/Helpers';
import Helptext from 'Components/Helptext';
import isEmpty from 'Libs/isEmpty';

import styles from './SeasonList.style';

type Props = {
  /* parent */
  tvshowId: string,
  seasons: Seasons,
  navigation: Object,
  /* connect */
  seasonViewings: { [season: string]: ViewerInfo[] },
  friends: Friends,
  viewingUpdate: Function,
  viewingUnview: Function,
  /* HOC */
  handleViewerPress: Function,
  handleViewerLongPress: Function,
  handleSeasonPress: Function,
};

const enhance = compose(
  withHandlers({
    handleViewerPress: ({ navigation }: Props) => (viewer: ViewerInfo) => {
      navigation.navigate('FriendDetailsPage', { friendId: viewer.friendId });
    },
    handleViewerLongPress: ({ tvshowId, viewingUnview }: Props) => (
      viewer: ViewerInfo,
    ) => {
      viewingUnview(tvshowId, viewer.friendId);
    },
    handleSeasonPress: ({ navigation, tvshowId, friends, viewingUpdate }: Props) => (
      seasonId: string,
    ) => {
      navigation.navigate('PickPage', {
        title: 'Who watched that season?',
        isMultiSelection: false,
        collection: friends,
        onSelect: friendId => {
          /* istanbul ignore next */
          viewingUpdate(tvshowId, friendId, seasonId);
        },
      });
    },
  }),
  pure,
);

const getBadgeBgColor = (color: string) => ({ backgroundColor: color });

const getBadgeColor = (color: string) => ({ color: getContrastingTextColor(color) });

function SeasonList({
  seasons,
  seasonViewings,
  handleViewerPress,
  handleViewerLongPress,
  handleSeasonPress,
}: Props) {
  let helpAlreadyDisplayed: boolean = false;
  const helpIfNeeded = () => {
    if (!helpAlreadyDisplayed && isEmpty(seasonViewings)) {
      helpAlreadyDisplayed = true;
      return <Helptext>Click on a season when someone viewed it</Helptext>;
    }
    return null;
  };

  return (
    <List>
      {seasons &&
        objectValues(seasons).reverse().map(season =>
          (<ListItem key={season.id} onPress={() => handleSeasonPress(season.id)}>
            <Grid>
              <Col size={Metrics.columnLeft}>
                <View style={styles.seasonHeader}>
                  {/* <Left style={styles.seasonButton}> */}
                  {/* <Icon name="add-circle" style={styles.icon} /> */}
                  {/* </Left> */}
                  <Text style={styles.seasonTitle}>{`Season ${season.id}`}</Text>
                  <Text note style={styles.seasonInfos}>
                    {season.episodes && `${season.episodes} ep.`}
                    {season.year && ` in ${season.year}`}
                  </Text>
                </View>
              </Col>
              <Col size={Metrics.columnRight} style={styles.viewerList}>
                {helpIfNeeded()}
                {seasonViewings[season.id] &&
                  seasonViewings[season.id].map(viewer =>
                    (<Button
                      small
                      key={viewer.friendId}
                      style={{ ...styles.viewer, ...getBadgeBgColor(viewer.color) }}
                      onPress={() => handleViewerPress(viewer)}
                      onLongPress={() => handleViewerLongPress(viewer)}
                    >
                      <Text style={getBadgeColor(viewer.color)}>
                        {viewer.name}
                      </Text>
                    </Button>),
                  )}
              </Col>
            </Grid>
          </ListItem>),
        )}
    </List>
  );
}

export default enhance(SeasonList);
