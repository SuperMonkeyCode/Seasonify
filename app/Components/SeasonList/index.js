/* @flow */
import { connect } from 'react-redux';

import { friendSelectors } from 'State/friendState';
import { viewingActions, viewingSelectors } from 'State/viewingState';

import SeasonList from './SeasonList';

export default connect(
  (state, { tvshowId }) => ({
    seasonViewings: viewingSelectors.getSeasonsWithViewers(state, { tvshowId }),
    friends: friendSelectors.getFriends(state),
  }),
  {
    viewingUpdate: viewingActions.viewingUpdate,
  },
)(SeasonList);
