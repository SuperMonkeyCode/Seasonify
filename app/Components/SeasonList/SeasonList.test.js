import { List } from 'native-base';
import { TouchableOpacity } from 'react-native';
import React from 'react';

import SeasonList from './SeasonList';

function setup(specificProps = {}) {
  const props = {
    navigation: NavigationMock,
    tvshowId: 's1',
    seasons: {
      1: {
        id: 1,
        episodes: 13,
        year: 2014,
      },
      2: {
        id: 2,
        episodes: 13,
        year: 2015,
      },
      3: {
        id: 3,
      },
    },
    seasonViewings: {
      2: [
        {
          friendId: 'f1',
          name: 'Friend 1',
          color: 'pink',
        },
        {
          friendId: 'f2',
          name: 'Friend 2',
          color: 'blue',
        },
      ],
    },
    viewingUnview: jest.fn(),
    ...specificProps,
  };
  const component = shallowDive(<SeasonList {...props} />, List);
  return {
    component,
    props,
  };
}

describe('Rendering', () => {
  it('should render', () => {
    const { component } = setup();
    expect(component).toMatchSnapshot();
  });
});

/* ========= Events & Functions ========= */

describe('Events & Functions', () => {
  const { component, props } = setup();
  const seasonHeader = component.find(TouchableOpacity).first();
  const buttonFriend1 = component.findWhere(node => node.key() === 'f1');

  describe('when calling onPress on Button', () => {
    buttonFriend1.props().onPress();

    it("navigates to friend's page", () => {
      expect(props.navigation.navigate).toBeCalledWith('FriendDetailsPage', {
        friendId: 'f1',
        friendName: 'Friend 1',
      });
    });
  });

  describe('when calling onLongPress on Button', () => {
    buttonFriend1.props().onLongPress();

    it('unview', () => {
      expect(props.viewingUnview).toBeCalledWith(props.tvshowId, 'f1');
    });
  });

  describe('when calling onPress on TouchableOpacity', () => {
    seasonHeader.props().onPress(1);

    it('navigates', () => {
      expect(props.navigation.navigate).toBeCalled();
    });
  });
});
