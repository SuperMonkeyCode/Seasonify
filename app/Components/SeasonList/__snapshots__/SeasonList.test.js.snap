// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering should render 1`] = `
<Styled(List)>
  <Styled(ListItem)>
    <GridNB>
      <ColumnNB
        size={38}
      >
        <TouchableOpacity
          activeOpacity={0.2}
          focusedOpacity={0.7}
          onPress={[Function]}
          style={
            Object {
              "alignItems": "flex-start",
              "flexDirection": "row",
              "flexWrap": "wrap",
              "justifyContent": "space-between",
            }
          }
          transparent={true}
        >
          <Styled(Left)
            style={Object {}}
          >
            <Styled(Icon)
              name="add-circle"
            />
          </Styled(Left)>
          <Styled(Text)
            style={
              Object {
                "flexGrow": 1,
                "textAlign": "right",
              }
            }
          >
            Season 3
          </Styled(Text)>
          <Styled(Text)
            note={true}
            style={
              Object {
                "flexBasis": "100%",
                "textAlign": "right",
              }
            }
          />
        </TouchableOpacity>
      </ColumnNB>
      <ColumnNB
        size={62}
        style={
          Object {
            "alignItems": "flex-start",
            "flexDirection": "row",
            "flexWrap": "wrap",
            "justifyContent": "flex-start",
            "paddingHorizontal": 20,
          }
        }
      />
    </GridNB>
  </Styled(ListItem)>
  <Styled(ListItem)>
    <GridNB>
      <ColumnNB
        size={38}
      >
        <TouchableOpacity
          activeOpacity={0.2}
          focusedOpacity={0.7}
          onPress={[Function]}
          style={
            Object {
              "alignItems": "flex-start",
              "flexDirection": "row",
              "flexWrap": "wrap",
              "justifyContent": "space-between",
            }
          }
          transparent={true}
        >
          <Styled(Left)
            style={Object {}}
          >
            <Styled(Icon)
              name="add-circle"
            />
          </Styled(Left)>
          <Styled(Text)
            style={
              Object {
                "flexGrow": 1,
                "textAlign": "right",
              }
            }
          >
            Season 2
          </Styled(Text)>
          <Styled(Text)
            note={true}
            style={
              Object {
                "flexBasis": "100%",
                "textAlign": "right",
              }
            }
          >
            13 ep.
             in 2015
          </Styled(Text)>
        </TouchableOpacity>
      </ColumnNB>
      <ColumnNB
        size={62}
        style={
          Object {
            "alignItems": "flex-start",
            "flexDirection": "row",
            "flexWrap": "wrap",
            "justifyContent": "flex-start",
            "paddingHorizontal": 20,
          }
        }
      >
        <Styled(Button)
          onLongPress={[Function]}
          onPress={[Function]}
          small={true}
          style={
            Object {
              "alignSelf": "flex-start",
              "backgroundColor": "pink",
              "marginHorizontal": 2,
              "marginVertical": 2,
            }
          }
        >
          <Styled(Text)
            style={
              Object {
                "color": "#ffffff",
              }
            }
          >
            Friend 1
          </Styled(Text)>
        </Styled(Button)>
        <Styled(Button)
          onLongPress={[Function]}
          onPress={[Function]}
          small={true}
          style={
            Object {
              "alignSelf": "flex-start",
              "backgroundColor": "blue",
              "marginHorizontal": 2,
              "marginVertical": 2,
            }
          }
        >
          <Styled(Text)
            style={
              Object {
                "color": "#ffffff",
              }
            }
          >
            Friend 2
          </Styled(Text)>
        </Styled(Button)>
      </ColumnNB>
    </GridNB>
  </Styled(ListItem)>
  <Styled(ListItem)>
    <GridNB>
      <ColumnNB
        size={38}
      >
        <TouchableOpacity
          activeOpacity={0.2}
          focusedOpacity={0.7}
          onPress={[Function]}
          style={
            Object {
              "alignItems": "flex-start",
              "flexDirection": "row",
              "flexWrap": "wrap",
              "justifyContent": "space-between",
            }
          }
          transparent={true}
        >
          <Styled(Left)
            style={Object {}}
          >
            <Styled(Icon)
              name="add-circle"
            />
          </Styled(Left)>
          <Styled(Text)
            style={
              Object {
                "flexGrow": 1,
                "textAlign": "right",
              }
            }
          >
            Season 1
          </Styled(Text)>
          <Styled(Text)
            note={true}
            style={
              Object {
                "flexBasis": "100%",
                "textAlign": "right",
              }
            }
          >
            13 ep.
             in 2014
          </Styled(Text)>
        </TouchableOpacity>
      </ColumnNB>
      <ColumnNB
        size={62}
        style={
          Object {
            "alignItems": "flex-start",
            "flexDirection": "row",
            "flexWrap": "wrap",
            "justifyContent": "flex-start",
            "paddingHorizontal": 20,
          }
        }
      />
    </GridNB>
  </Styled(ListItem)>
</Styled(List)>
`;
